File: src/codec_g723_1.c
#include "codec_g723_1.h"
#include "g723_backend.h"
#include "backends/c/g723_resample.h"

#include <stddef.h>
#include <stdint.h>
#include <stdlib.h>

typedef struct {
    const g723_backend_v1_t *backend;
    g723_backend_encoder_t *backend_encoder;
    uint32_t sample_rate;
    g723_resampler_state_t resampler;
} g723_encoder_instance_t;

typedef struct {
    const g723_backend_v1_t *backend;
    g723_backend_decoder_t *backend_decoder;
    uint32_t sample_rate;
    g723_resampler_state_t resampler;
} g723_decoder_instance_t;

static const g723_backend_v1_t *get_backend(void) {
    const g723_backend_v1_t *backend =
        g723_get_compiled_backend();

    if (backend == NULL) {
        return NULL;
    }

    if (backend->abi_version != G723_ABI_VERSION) {
        return NULL;
    }

    return backend;
}

uint32_t g723_abi_version(void) {
    return G723_ABI_VERSION;
}

const char *g723_backend_name(void) {
    const g723_backend_v1_t *backend = get_backend();

    if (backend == NULL || backend->name == NULL) {
        return "unknown";
    }

    return backend->name;
}

int32_t g723_backend_supports_bitrate(int32_t bitrate) {
    const g723_backend_v1_t *backend = get_backend();

    if (backend == NULL ||
        backend->supports_bitrate == NULL) {
        return 0;
    }

    if (bitrate != G723_BITRATE_5300 &&
        bitrate != G723_BITRATE_6300) {
        return 0;
    }

    return backend->supports_bitrate(
        (g723_bitrate_t)bitrate
    ) ? 1 : 0;
}

int32_t g723_encoder_create(
    int32_t bitrate,
    void **encoder_out
) {
    const g723_backend_v1_t *backend;
    g723_backend_encoder_t *backend_encoder = NULL;
    g723_encoder_instance_t *instance;
    g723_result_t result;

    if (encoder_out == NULL) {
        return G723_RESULT_INVALID_ARGUMENT;
    }

    *encoder_out = NULL;

    if (bitrate != G723_BITRATE_5300 &&
        bitrate != G723_BITRATE_6300) {
        return G723_RESULT_INVALID_ARGUMENT;
    }

    backend = get_backend();

    if (backend == NULL ||
        backend->encoder_create == NULL ||
        backend->encoder_destroy == NULL) {
        return G723_RESULT_INTERNAL;
    }

    if (backend->supports_bitrate == NULL ||
        !backend->supports_bitrate(
            (g723_bitrate_t)bitrate
        )) {
        return G723_RESULT_UNSUPPORTED;
    }

    result = backend->encoder_create(
        (g723_bitrate_t)bitrate,
        &backend_encoder
    );

    if (result != G723_RESULT_OK) {
        return result;
    }

    if (backend_encoder == NULL) {
        return G723_RESULT_INTERNAL;
    }

    instance = (g723_encoder_instance_t *)
        malloc(sizeof(g723_encoder_instance_t));

    if (instance == NULL) {
        backend->encoder_destroy(backend_encoder);
        return G723_RESULT_NO_MEMORY;
    }

    instance->backend = backend;
    instance->backend_encoder = backend_encoder;
    instance->sample_rate = G723_PCM_SAMPLE_RATE;
    g723_resampler_init(&instance->resampler);

    *encoder_out = instance;

    return G723_RESULT_OK;
}

int32_t g723_encoder_set_sample_rate(void *encoder, uint32_t sample_rate) {
    if (encoder == NULL) return G723_RESULT_INVALID_ARGUMENT;
    if (sample_rate != G723_PCM_SAMPLE_RATE && sample_rate != G723_PCM_SAMPLE_RATE_16K) {
        return G723_RESULT_INVALID_ARGUMENT;
    }
    g723_encoder_instance_t *instance = (g723_encoder_instance_t *)encoder;
    instance->sample_rate = sample_rate;
    return G723_RESULT_OK;
}

void g723_encoder_destroy(void *encoder) {
    g723_encoder_instance_t *instance;

    if (encoder == NULL) {
        return;
    }

    instance = (g723_encoder_instance_t *)encoder;

    if (instance->backend != NULL &&
        instance->backend->encoder_destroy != NULL &&
        instance->backend_encoder != NULL) {
        instance->backend->encoder_destroy(
            instance->backend_encoder
        );
    }

    free(instance);
}

int32_t g723_encode_frame(
    void *encoder,
    const int16_t *pcm,
    uint32_t pcm_samples,
    uint8_t *output,
    uint32_t output_capacity,
    uint32_t *output_size
) {
    g723_encoder_instance_t *instance;
    size_t backend_output_size = 0;
    g723_result_t result;

    if (output_size != NULL) {
        *output_size = 0;
    }

    if (encoder == NULL ||
        pcm == NULL ||
        output == NULL ||
        output_size == NULL) {
        return G723_RESULT_INVALID_ARGUMENT;
    }

    if (pcm_samples != G723_PCM_SAMPLES_PER_FRAME &&
        pcm_samples != G723_PCM_SAMPLES_PER_FRAME_16K) {
        return G723_RESULT_INVALID_ARGUMENT;
    }

    if (output_capacity < G723_FRAME_BYTES_6300) {
        return G723_RESULT_INVALID_ARGUMENT;
    }

    instance = (g723_encoder_instance_t *)encoder;

    if (instance->backend == NULL ||
        instance->backend->encode_frame == NULL ||
        instance->backend_encoder == NULL) {
        return G723_RESULT_INTERNAL;
    }

    const int16_t *pcm_to_encode = pcm;
    int16_t pcm_8k[G723_PCM_SAMPLES_PER_FRAME];
    if (pcm_samples == G723_PCM_SAMPLES_PER_FRAME_16K) {
        g723_resample_16k_to_8k(&instance->resampler, pcm, pcm_8k);
        pcm_to_encode = pcm_8k;
    }

    result = instance->backend->encode_frame(
        instance->backend_encoder,
        pcm_to_encode,
        output,
        (size_t)output_capacity,
        &backend_output_size
    );

    if (result != G723_RESULT_OK) {
        return result;
    }

    if (backend_output_size > output_capacity ||
        backend_output_size > UINT32_MAX) {
        return G723_RESULT_INTERNAL;
    }

    *output_size = (uint32_t)backend_output_size;

    return G723_RESULT_OK;
}

int32_t g723_decoder_create(
    void **decoder_out
) {
    const g723_backend_v1_t *backend;
    g723_backend_decoder_t *backend_decoder = NULL;
    g723_decoder_instance_t *instance;
    g723_result_t result;

    if (decoder_out == NULL) {
        return G723_RESULT_INVALID_ARGUMENT;
    }

    *decoder_out = NULL;

    backend = get_backend();

    if (backend == NULL ||
        backend->decoder_create == NULL ||
        backend->decoder_destroy == NULL) {
        return G723_RESULT_INTERNAL;
    }

    result = backend->decoder_create(
        &backend_decoder
    );

    if (result != G723_RESULT_OK) {
        return result;
    }

    if (backend_decoder == NULL) {
        return G723_RESULT_INTERNAL;
    }

    instance = (g723_decoder_instance_t *)
        malloc(sizeof(g723_decoder_instance_t));

    if (instance == NULL) {
        backend->decoder_destroy(backend_decoder);
        return G723_RESULT_NO_MEMORY;
    }

    instance->backend = backend;
    instance->backend_decoder = backend_decoder;
    instance->sample_rate = G723_PCM_SAMPLE_RATE;
    g723_resampler_init(&instance->resampler);

    *decoder_out = instance;

    return G723_RESULT_OK;
}

int32_t g723_decoder_set_sample_rate(void *decoder, uint32_t sample_rate) {
    if (decoder == NULL) return G723_RESULT_INVALID_ARGUMENT;
    if (sample_rate != G723_PCM_SAMPLE_RATE && sample_rate != G723_PCM_SAMPLE_RATE_16K) {
        return G723_RESULT_INVALID_ARGUMENT;
    }
    g723_decoder_instance_t *instance = (g723_decoder_instance_t *)decoder;
    instance->sample_rate = sample_rate;
    return G723_RESULT_OK;
}

void g723_decoder_destroy(void *decoder) {
    g723_decoder_instance_t *instance;

    if (decoder == NULL) {
        return;
    }

    instance = (g723_decoder_instance_t *)decoder;

    if (instance->backend != NULL &&
        instance->backend->decoder_destroy != NULL &&
        instance->backend_decoder != NULL) {
        instance->backend->decoder_destroy(
            instance->backend_decoder
        );
    }

    free(instance);
}

int32_t g723_decode_frame(
    void *decoder,
    const uint8_t *input,
    uint32_t input_size,
    int16_t *pcm,
    uint32_t pcm_capacity,
    uint32_t *pcm_samples
) {
    g723_decoder_instance_t *instance;
    g723_result_t result;

    if (pcm_samples != NULL) {
        *pcm_samples = 0;
    }

    if (decoder == NULL ||
        input == NULL ||
        input_size == 0 ||
        pcm == NULL ||
        pcm_samples == NULL) {
        return G723_RESULT_INVALID_ARGUMENT;
    }

    instance = (g723_decoder_instance_t *)decoder;

    if (instance->backend == NULL ||
        instance->backend->decode_frame == NULL ||
        instance->backend_decoder == NULL) {
        return G723_RESULT_INTERNAL;
    }

    if (instance->sample_rate == G723_PCM_SAMPLE_RATE_16K) {
        if (pcm_capacity < G723_PCM_SAMPLES_PER_FRAME_16K) {
            return G723_RESULT_INVALID_ARGUMENT;
        }

        int16_t pcm_8k[G723_PCM_SAMPLES_PER_FRAME];
        result = instance->backend->decode_frame(
            instance->backend_decoder,
            input,
            (size_t)input_size,
            pcm_8k
        );

        if (result != G723_RESULT_OK) {
            return result;
        }

        g723_resample_8k_to_16k(&instance->resampler, pcm_8k, pcm);
        *pcm_samples = G723_PCM_SAMPLES_PER_FRAME_16K;
    } else {
        if (pcm_capacity < G723_PCM_SAMPLES_PER_FRAME) {
            return G723_RESULT_INVALID_ARGUMENT;
        }

        result = instance->backend->decode_frame(
            instance->backend_decoder,
            input,
            (size_t)input_size,
            pcm
        );

        if (result != G723_RESULT_OK) {
            return result;
        }

        *pcm_samples = G723_PCM_SAMPLES_PER_FRAME;
    }

    return G723_RESULT_OK;
}

int32_t g723_frame_type(
    const uint8_t *input,
    uint32_t input_size
) {
    const g723_backend_v1_t *backend = get_backend();

    if (input == NULL || input_size == 0) {
        return G723_FRAME_TYPE_INVALID;
    }

    if (backend == NULL ||
        backend->frame_type == NULL) {
        return G723_FRAME_TYPE_INVALID;
    }

    return (int32_t)backend->frame_type(
        input,
        (size_t)input_size
    );
}

uint32_t g723_frame_size(
    int32_t frame_type
) {
    const g723_backend_v1_t *backend = get_backend();
    size_t size;

    if (backend == NULL ||
        backend->frame_size == NULL) {
        return 0;
    }

    switch (frame_type) {
        case G723_FRAME_TYPE_6300:
        case G723_FRAME_TYPE_5300:
        case G723_FRAME_TYPE_SID:
        case G723_FRAME_TYPE_UNTRANSMITTED:
            break;

        default:
            return 0;
    }

    size = backend->frame_size(
        (g723_frame_type_t)frame_type
    );

    if (size > UINT32_MAX) {
        return 0;
    }

    return (uint32_t)size;
}

File: lib/codec_g723_1.dart
library;

export 'src/converter.dart';
export 'src/wav_file.dart';

import 'dart:ffi' as ffi;
import 'dart:typed_data';

import 'package:ffi/ffi.dart';

import 'codec_g723_1_bindings_generated.dart' as native;

enum G723Bitrate {
  kbps53(5300),
  kbps63(6300);

  const G723Bitrate(this.bitsPerSecond);

  final int bitsPerSecond;
}

enum G723FrameType { kbps63, kbps53, sid, untransmitted, invalid }

class G723CodecException implements Exception {
  const G723CodecException(this.message, {this.nativeCode});

  final String message;
  final int? nativeCode;

  @override
  String toString() {
    if (nativeCode == null) {
      return 'G723CodecException: $message';
    }

    return 'G723CodecException: $message '
        '(native error $nativeCode)';
  }
}

class G723UnsupportedException extends G723CodecException {
  const G723UnsupportedException(super.message)
    : super(nativeCode: -2);
}

class G723CodecInfo {
  const G723CodecInfo({
    required this.abiVersion,
    required this.backendName,
    required this.supports53k,
    required this.supports63k,
  });

  final int abiVersion;
  final String backendName;
  final bool supports53k;
  final bool supports63k;
}

abstract final class G723Codec {
  static const int samplesPerFrame = 240;
  static const int sampleRate = 8000;

  static const int samplesPerFrame16k = 480;
  static const int sampleRate16k = 16000;

  static const int frameBytes53 = 20;
  static const int frameBytes63 = 24;
  static const int frameBytesSid = 4;
  static const int frameBytesUntransmitted = 1;

  static int get abiVersion => native.g723_abi_version();

  static String get backendName {
    final pointer = native.g723_backend_name();

    if (pointer == ffi.nullptr) {
      return 'unknown';
    }

    return pointer.cast<Utf8>().toDartString();
  }

  static bool supportsBitrate(G723Bitrate bitrate) {
    return native.g723_backend_supports_bitrate(bitrate.bitsPerSecond) != 0;
  }

  static G723CodecInfo get info {
    return G723CodecInfo(
      abiVersion: abiVersion,
      backendName: backendName,
      supports53k: supportsBitrate(G723Bitrate.kbps53),
      supports63k: supportsBitrate(G723Bitrate.kbps63),
    );
  }

  static G723FrameType frameType(Uint8List frame) {
    if (frame.isEmpty) {
      return G723FrameType.invalid;
    }

    final input = calloc<ffi.Uint8>(frame.length);

    try {
      input.asTypedList(frame.length).setAll(0, frame);

      final value = native.g723_frame_type(input, frame.length);

      return switch (value) {
        0 => G723FrameType.kbps63,
        1 => G723FrameType.kbps53,
        2 => G723FrameType.sid,
        3 => G723FrameType.untransmitted,
        _ => G723FrameType.invalid,
      };
    } finally {
      calloc.free(input);
    }
  }

  static int expectedFrameSize(G723FrameType type) {
    final nativeType = switch (type) {
      G723FrameType.kbps63 => 0,
      G723FrameType.kbps53 => 1,
      G723FrameType.sid => 2,
      G723FrameType.untransmitted => 3,
      G723FrameType.invalid => 255,
    };

    return native.g723_frame_size(nativeType);
  }
}

class G723Encoder {
  G723Encoder({
    required this.bitrate,
    int sampleRate = G723Codec.sampleRate,
  }) : _sampleRate = sampleRate {
    if (sampleRate != G723Codec.sampleRate &&
        sampleRate != G723Codec.sampleRate16k) {
      throw ArgumentError.value(
        sampleRate,
        'sampleRate',
        'Sample rate must be either ${G723Codec.sampleRate} or '
            '${G723Codec.sampleRate16k} Hz.',
      );
    }

    final out = calloc<ffi.Pointer<ffi.Void>>();

    try {
      final result = native.g723_encoder_create(bitrate.bitsPerSecond, out);

      _throwIfError(result, operation: 'Creating G.723.1 encoder');

      if (out.value == ffi.nullptr) {
        throw const G723CodecException(
          'Native encoder creation returned a null handle.',
        );
      }

      _handle = out.value;

      if (sampleRate != G723Codec.sampleRate) {
        final srResult = native.g723_encoder_set_sample_rate(
          _handle,
          sampleRate,
        );
        _throwIfError(srResult, operation: 'Setting G.723.1 encoder sample rate');
      }
    } finally {
      calloc.free(out);
    }
  }

  final G723Bitrate bitrate;
  int _sampleRate;

  int get sampleRate => _sampleRate;

  int get samplesPerFrame => _sampleRate == G723Codec.sampleRate16k
      ? G723Codec.samplesPerFrame16k
      : G723Codec.samplesPerFrame;

  ffi.Pointer<ffi.Void> _handle = ffi.nullptr;
  bool _closed = false;

  bool get isClosed => _closed;

  void setSampleRate(int rate) {
    _checkOpen();

    if (rate != G723Codec.sampleRate && rate != G723Codec.sampleRate16k) {
      throw ArgumentError.value(
        rate,
        'rate',
        'Sample rate must be either ${G723Codec.sampleRate} or '
            '${G723Codec.sampleRate16k} Hz.',
      );
    }

    final result = native.g723_encoder_set_sample_rate(_handle, rate);
    _throwIfError(result, operation: 'Setting G.723.1 encoder sample rate');
    _sampleRate = rate;
  }

  Uint8List encodeFrame(Int16List pcm) {
    _checkOpen();

    final expectedSamples = samplesPerFrame;

    if (pcm.length != expectedSamples) {
      throw ArgumentError.value(
        pcm.length,
        'pcm.length',
        'A G.723.1 frame must contain exactly '
            '$expectedSamples samples at $_sampleRate Hz.',
      );
    }

    final nativePcm = calloc<ffi.Int16>(expectedSamples);

    final output = calloc<ffi.Uint8>(G723Codec.frameBytes63);

    final outputSize = calloc<ffi.Uint32>();

    try {
      nativePcm.asTypedList(expectedSamples).setAll(0, pcm);

      final result = native.g723_encode_frame(
        _handle,
        nativePcm,
        expectedSamples,
        output,
        G723Codec.frameBytes63,
        outputSize,
      );

      _throwIfError(result, operation: 'Encoding G.723.1 frame');

      final size = outputSize.value;

      if (size > G723Codec.frameBytes63) {
        throw G723CodecException(
          'Native encoder returned invalid frame size $size.',
        );
      }

      return Uint8List.fromList(output.asTypedList(size));
    } finally {
      calloc.free(outputSize);
      calloc.free(output);
      calloc.free(nativePcm);
    }
  }

  void close() {
    if (_closed) {
      return;
    }

    if (_handle != ffi.nullptr) {
      native.g723_encoder_destroy(_handle);
      _handle = ffi.nullptr;
    }

    _closed = true;
  }

  void _checkOpen() {
    if (_closed || _handle == ffi.nullptr) {
      throw StateError('The G.723.1 encoder has already been closed.');
    }
  }
}

class G723Decoder {
  G723Decoder({int sampleRate = G723Codec.sampleRate})
      : _sampleRate = sampleRate {
    if (sampleRate != G723Codec.sampleRate &&
        sampleRate != G723Codec.sampleRate16k) {
      throw ArgumentError.value(
        sampleRate,
        'sampleRate',
        'Sample rate must be either ${G723Codec.sampleRate} or '
            '${G723Codec.sampleRate16k} Hz.',
      );
    }

    final out = calloc<ffi.Pointer<ffi.Void>>();

    try {
      final result = native.g723_decoder_create(out);

      _throwIfError(result, operation: 'Creating G.723.1 decoder');

      if (out.value == ffi.nullptr) {
        throw const G723CodecException(
          'Native decoder creation returned a null handle.',
        );
      }

      _handle = out.value;

      if (sampleRate != G723Codec.sampleRate) {
        final srResult = native.g723_decoder_set_sample_rate(
          _handle,
          sampleRate,
        );
        _throwIfError(srResult, operation: 'Setting G.723.1 decoder sample rate');
      }
    } finally {
      calloc.free(out);
    }
  }

  int _sampleRate;

  int get sampleRate => _sampleRate;

  int get samplesPerFrame => _sampleRate == G723Codec.sampleRate16k
      ? G723Codec.samplesPerFrame16k
      : G723Codec.samplesPerFrame;

  ffi.Pointer<ffi.Void> _handle = ffi.nullptr;
  bool _closed = false;

  bool get isClosed => _closed;

  void setSampleRate(int rate) {
    _checkOpen();

    if (rate != G723Codec.sampleRate && rate != G723Codec.sampleRate16k) {
      throw ArgumentError.value(
        rate,
        'rate',
        'Sample rate must be either ${G723Codec.sampleRate} or '
            '${G723Codec.sampleRate16k} Hz.',
      );
    }

    final result = native.g723_decoder_set_sample_rate(_handle, rate);
    _throwIfError(result, operation: 'Setting G.723.1 decoder sample rate');
    _sampleRate = rate;
  }

  Int16List decodeFrame(Uint8List frame) {
    _checkOpen();

    if (frame.isEmpty) {
      throw ArgumentError.value(
        frame.length,
        'frame.length',
        'Encoded frame must not be empty.',
      );
    }

    final type = G723Codec.frameType(frame);
    final expectedSize = G723Codec.expectedFrameSize(type);

    if (type == G723FrameType.invalid || expectedSize == 0) {
      throw const G723CodecException('Invalid G.723.1 frame type.');
    }

    if (frame.length != expectedSize) {
      throw G723CodecException(
        'Invalid G.723.1 frame size: '
        'got ${frame.length}, expected $expectedSize.',
      );
    }

    final expectedSamples = samplesPerFrame;
    final input = calloc<ffi.Uint8>(frame.length);
    final pcm = calloc<ffi.Int16>(expectedSamples);
    final pcmSamples = calloc<ffi.Uint32>();

    try {
      input.asTypedList(frame.length).setAll(0, frame);

      final result = native.g723_decode_frame(
        _handle,
        input,
        frame.length,
        pcm,
        expectedSamples,
        pcmSamples,
      );

      _throwIfError(result, operation: 'Decoding G.723.1 frame');

      final count = pcmSamples.value;

      if (count > expectedSamples) {
        throw G723CodecException(
          'Native decoder returned invalid PCM '
          'sample count $count.',
        );
      }

      return Int16List.fromList(pcm.asTypedList(count));
    } finally {
      calloc.free(pcmSamples);
      calloc.free(pcm);
      calloc.free(input);
    }
  }

  void close() {
    if (_closed) {
      return;
    }

    if (_handle != ffi.nullptr) {
      native.g723_decoder_destroy(_handle);
      _handle = ffi.nullptr;
    }

    _closed = true;
  }

  void _checkOpen() {
    if (_closed || _handle == ffi.nullptr) {
      throw StateError('The G.723.1 decoder has already been closed.');
    }
  }
}

void _throwIfError(int result, {required String operation}) {
  switch (result) {
    case 0:
      return;

    case -1:
      throw G723CodecException(
        '$operation failed: invalid argument.',
        nativeCode: result,
      );

    case -2:
      throw G723UnsupportedException(
        '$operation is not supported by the '
        'current codec backend.',
      );

    case -3:
      throw G723CodecException(
        '$operation failed: bad G.723.1 frame.',
        nativeCode: result,
      );

    case -4:
      throw G723CodecException(
        '$operation failed: native memory allocation failed.',
        nativeCode: result,
      );

    case -5:
      throw G723CodecException(
        '$operation failed: internal native error.',
        nativeCode: result,
      );

    default:
      throw G723CodecException(
        '$operation failed with an unknown native error.',
        nativeCode: result,
      );
  }
}
